@(iteration: Int, cluster: Int, rows: List[String], columns: List[String],
  ratios: RatioMatrix, motifInfoMap: Map[String, Array[MotifInfo]],
  pssmMap: Map[String, Array[String]],
  annotationMap: Map[String, Seq[GeneAnnotations]],
  gagglePSSMs: Map[String, Array[String]])

<div id="cluster-view">
  <script type="text/javascript">
    @motifInfoMap.keys.map { seqType =>
      @for(i <- 0 until pssmMap(seqType).length) {
        var pssm@(seqType)_@i = @Html(pssmMap(seqType)(i));
      }
    }
    var canvasOptions = {
        width: 300,
        height: 150,
        glyphStyle: '20pt Helvetica'
    };

    $(document).ready(function() {
        var chart1 = new Highcharts.Chart({
            chart: {
                renderTo: 'expression-graph',
                type: 'line'
            },
            title: {
                text: 'Gene Expressions'
            },
            plotOptions: { line: { marker: { enabled: false } } },
            yAxis: {
                title: {
                    text: 'log10'
                }
            },
            series: @(HighchartsFormatter.toHSSeries(ratios))
        });
        @motifInfoMap.keys.map { seqType =>
          @for(i <- 0 until pssmMap(seqType).length) {
            isblogo.makeLogo('canvas@(seqType)_@i', pssm@(seqType)_@i, canvasOptions);
          }
        }
        @motifInfoMap.keys.map { seqType =>
        var annotations_@seqType = [
            @for(i <- 0 until annotationMap(seqType).length) {

            {
                gene: '@(annotationMap(seqType)(i).gene)', condition: '',
                log10: -0.17,
                boxColor: '#08f',
                lineColor: '#000',
                matches: [
                    @for(j <- 0 until annotationMap(seqType)(i).annotations.length) {
                        @defining(annotationMap(seqType)(i).annotations(j)) { annotation =>
                                                                   { motif: @(annotation.motifNum - 1), start: @(annotation.position), length: @motifInfoMap(seqType)(annotation.motifNum - 1).pssm.length, reverse: @(annotation.reverse), score: @(annotation.pvalue) },
                                                                 }
                    }
                ]
            },

            }
        ];
        annot.draw('annotation-graph_@seqType', { width: 500, height: 400, annotations: annotations_@seqType});
        }
    });
  </script>
  <h2>Cluster @cluster at iteration @iteration</h2>
  <h3>Gene expressions</h3>
  <div id="expression-graph"></div>
  <div>&nbsp;</div>
  <h3>Motifs</h3>
  @motifInfoMap.keys.map { seqType =>
  @for(i <- 0 until pssmMap(seqType).length) {
    <span style="display: inline; float: left; margin-right: 10px;">
    <div>Sequence type: @seqType, motif @i, evalue: @(motifInfoMap(seqType)(i).evalue)</div>
    <div id="canvas@(seqType)_@i"></div>
    </span>
  }
  }
  <div style="clear: both"></div>
  @motifInfoMap.keys.map { seqType =>
  <h3>Annotations for @seqType</h3>
  <div id="annotation-graph_@seqType"></div>
  }
  <div style="clear: both"></div>
  <div>&nbsp;</div>
  <h3>Cluster memberships</h3>
  <div id="cluster-members">
    <span style="float: left; margin-right: 20px">
      <span>Row members</span>
      <table class="members">
        <tr><th>Gene</th></th>
        @for(row <- rows) {<tr><td>@row</td></tr>
        }
      </table>
    </span>
    <span>
      <span>Column members</span>
      <table class="members">
        <tr><th>Condition</th></th>
        @for(col <- columns) {<tr><td>@col</td></tr>
        }
      </table>
    </span>
  </div>
</div>

@motifInfoMap.keys.map { seqType =>
  @for(i <- 0 until gagglePSSMs(seqType).length) {
    <div class="gaggle-matrix-tsv">
      @(gagglePSSMs(seqType)(i))
    </div>
  }
}
